modelhub/modelhub/models/funnel_discovery.py

```diff
@@ -0,0 +1,456 @@
+            )
+        elif is_postgres(data.engine):
+            flattened_lc = flattened_lc.copy_override(
+                expression=bach.expression.Expression.construct(
+                    '{} #>> {}',
+                    flattened_lc,
+                    bach.expression.Expression.raw("'{}'")
+                )
+            )
+
+        flattened_lc = flattened_lc.astype('string')
+
+        offset_lc = offset_lc.copy_override(name='__root_step_offset')
+
+        nav_df = flattened_lc.to_frame()
+        nav_df[offset_lc.name] = offset_lc
+
+        nav_df = nav_df.sort_values(
+            by=nav_df.index_columns + [offset_lc.name],
+            # since we are using lag, we should reverse the step order
+            ascending=[True] * len(nav_df.index_columns) + [False],
+        )
+
+        window = nav_df.groupby(by=nav_df.index_columns).window()
+        root_step_series = window[flattened_lc.name]
+
+        all_step_series = {}
+        for step in range(1, steps + 1):
+            step_series_name = f'{flattened_lc.name}_step_{step}'
+            if step == 1:
+                next_step = root_step_series.copy_override(group_by=None)
+            else:
+                next_step = root_step_series.window_lag(offset=step - 1)
+
+            all_step_series[step_series_name] = (
+                next_step.copy_override(name=step_series_name)
+            )
+
+        result = nav_df.copy_override(
+            base_node=window.base_node,
+            series={
+                **all_step_series,
+                offset_lc.name: offset_lc.copy_override(base_node=window.base_node),
+            }
+        )
+        result = result.materialize(node_name='step_extraction')
+
+        # limit rows
+        if n_examples is not None:
+            result = result[result[offset_lc.name] < n_examples]
+
+        # re-order rows
+        result = result.sort_values(by=result.index_columns + [offset_lc.name])
+        result = result.drop(columns=[offset_lc.name])
+
+        # conversion part
+        if add_conversion_step_column or only_converted_paths:
+            if 'feature_nice_name' not in data.data_columns:
+                data['feature_nice_name'] = data.location_stack.ls.nice_name
+
+            if add_conversion_step_column:
+                result = self._add_first_conversion_step_number_column(result,
+                                                                       data)
+            if only_converted_paths:
+                if self.conversion_step_column not in result.data_columns:
+                    result = self._add_first_conversion_step_number_column(result,
+                                                                           data)
+                result = self._filter_navigation_paths_to_conversion(result)
+                if add_conversion_step_column is False:
+                    result = result.drop(columns=[self.conversion_step_column])
```

Comment:
I think this is a bit more clear
```suggestion
        if not (add_conversion_step_column or only_converted_paths):
            return result
            
        if 'feature_nice_name' not in data.data_columns:
            data['feature_nice_name'] = data.location_stack.ls.nice_name

        result = self._add_first_conversion_step_number_column(result, data)
        
        if not only_converted_paths:
            return result
        
        result = self._filter_navigation_paths_to_conversion(result)
        
        if add_conversion_step_column is False:
            result = result.drop(columns=[self.conversion_step_column])
                
        return result
```